modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/idempotency/NuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,168 @@
+
+    protected static final int DEFERRED_OUTPUT_STREAM_THRESHOLD = 1024 * 1024; // 1 MB
+
+    // safe methods according to RFC 7231 4.2.1
+    protected static final Set<String> SAFE_METHODS = Set.of(HttpGet.METHOD_NAME, HttpHead.METHOD_NAME,
+            HttpOptions.METHOD_NAME, HttpTrace.METHOD_NAME);
+
+    // idempotent methods according to RFC 7231 4.2.2
+    protected static final Set<String> IDEMPOTENT_METHODS = Set.of(HttpPut.METHOD_NAME, HttpDelete.METHOD_NAME);
+
+    protected long getTTLSeconds() {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        if (cs != null) {
+            return cs.getLong(TTL_SECONDS_PROPERTY, DEFAULT_TTL_SECONDS);
+        }
+        return DEFAULT_TTL_SECONDS;
+    }
+
+    protected String getStoreName() {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        if (cs != null) {
+            return cs.getString(STORE_PROPERTY, DEFAULT_STORE);
+        }
+        return DEFAULT_STORE;
+    }
+
+    protected boolean isIdempotentMethod(String method) {
+        return SAFE_METHODS.contains(method) || IDEMPOTENT_METHODS.contains(method);
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+        String method = request.getMethod();
+        if (isIdempotentMethod(method)) {
+            log.debug("No idempotent processing done: method is already idempotent: " + method);
+            chain.doFilter(request, response);
+            return;
+        }
+        String key = request.getHeader(NuxeoIdempotentFilter.HEADER_KEY);
+        if (key == null) {
+            log.debug("No idempotent processing done: no Idempotent-Key present");
+            chain.doFilter(request, response);
+            return;
+        }
+        log.debug("Idempotent request key: " + key);
+        KeyValueService kvs = Framework.getService(KeyValueService.class);
+        if (kvs == null) {
+            log.debug("KeyValueService not present");
+            chain.doFilter(request, response);
+            return;
+        }
+        KeyValueStore store = kvs.getKeyValueStore(getStoreName());
+        String status = store.getString(key + NuxeoIdempotentFilter.INFO_SUFFIX);
+        if (status == null) {
+            log.debug("Handle new request for key: " + key);
+            long ttl = getTTLSeconds();
+            store.put(key + NuxeoIdempotentFilter.INFO_SUFFIX, NuxeoIdempotentFilter.INPROGRESS_MARKER, ttl);
+            try {
+                CopyingResponseWrapper wrapper = new CopyingResponseWrapper(DEFERRED_OUTPUT_STREAM_THRESHOLD, response);
+                chain.doFilter(request, wrapper);
+                store.put(key, wrapper.getCaptureAsBytes(), ttl);
```

Comment:
We should have a protection here against captures that are too big, as the k/v store is not designed to store arbitrarily big payloads.